crypto/evp/pmeth_lib.c

```diff
@@ -734,6 +734,35 @@ static int legacy_ctrl_to_param(EVP_PKEY_CTX *ctx, int keytype, int optype,
 # ifndef OPENSSL_NO_DH
     case EVP_PKEY_CTRL_DH_PAD:
         return EVP_PKEY_CTX_set_dh_pad(ctx, p1);
+# endif
+# ifndef OPENSSL_NO_EC
+    case EVP_PKEY_CTRL_EC_ECDH_COFACTOR:
+        if (p1 == -2) {
+            return EVP_PKEY_CTX_get_ecdh_cofactor_mode(ctx);
+        } else if (p1 < -1 || p1 > 1) {
+            /* Uses the same return values as EVP_PKEY_CTX_ctrl */
+            return -2;
+        } else {
+            return EVP_PKEY_CTX_set_ecdh_cofactor_mode(ctx, p1);
+        }
+    case EVP_PKEY_CTRL_EC_KDF_TYPE:
+        if (p1 == -2) {
+            return EVP_PKEY_CTX_get_ecdh_kdf_type(ctx);
+        } else {
+            return EVP_PKEY_CTX_set_ecdh_kdf_type(ctx, p1);
+        }
+    case EVP_PKEY_CTRL_GET_EC_KDF_MD:
+        return EVP_PKEY_CTX_get_ecdh_kdf_md(ctx, p2);
+    case EVP_PKEY_CTRL_EC_KDF_MD:
+        return EVP_PKEY_CTX_set_ecdh_kdf_md(ctx, p2);
+    case EVP_PKEY_CTRL_GET_EC_KDF_OUTLEN:
+        return EVP_PKEY_CTX_get_ecdh_kdf_outlen(ctx, p2);
+    case EVP_PKEY_CTRL_EC_KDF_OUTLEN:
+        return EVP_PKEY_CTX_set_ecdh_kdf_outlen(ctx, p1);
+    case EVP_PKEY_CTRL_GET_EC_KDF_UKM:
+        return EVP_PKEY_CTX_get0_ecdh_kdf_ukm(ctx, p2);
+    case EVP_PKEY_CTRL_EC_KDF_UKM:
+        return EVP_PKEY_CTX_set0_ecdh_kdf_ukm(ctx, p2, p1);
```

Comment:
The cases here clash between cryptosystems: e.g. `EVP_PKEY_CTRL_GET_EC_KDF_MD` could be using the same numeric value as `EVP_PKEY_SET_RSA_FOO` (don't remember which case clashed with which, but most of them were overlapping before the hack below).

What is our plan to solve this?

Reply:
As per our discussion I think you should first try putting these switch statements in an "if" to test the key type to avoid clashes.